### PR TITLE
enhancement(datadog_logs sink): support datadog logs new HTTPS transport

### LINKED ISSUE #2504: Support Datadog Log's new HTTPs transport
In the latest Datadog Agent 6.19 and 7.19 new https transport is enforced by default for sending logs. Introduced in 6.14 and 7.14 and if any problems falling back to TCP transport.

This option is very nice because it allows standard HTTP compression on the transport layer which saves a lot of bandwidth on a network. 
Also, introduce some batching/rate limit conditions.

https://docs.datadoghq.com/agent/logs/log_transport/?tab=https

I think this. should be included in Vector and somehow similar to that one from Datadog metrics sink.

### PR DESCRIPTION
**Pull request by FungusHumungus**
This fixes #2504 

Updated the sink to use Https rather that TCP.

I had to create 2 separate services to support both JSON and Text as the batching policies were different for each. I couldn't find any precedent for this in any other services, so it is possible I have missed something and there is a better way.

Note this is potentially a breaking change because it now uses the Datadogs http endpoint. If a user has specified a specific Tcp endpoint in their config, this will likely no longer work. Also, any TLS settings are now obsolete.

### CODE REVIEW

**Comment by binarylogic** [comment]
Assuming this is significantly better from a performance/bandwidth standpoint, and Datadog is pushing for this, I think it's fine to flag this as a breaking change.

**Comment by jamtur01** [comment]
@FungusHumungus Is there an still an issue with the region? Do we need to make it configurable still?

**Comment by FungusHumungus** [comment]
@jamtur01 Yes there is. I've almost got that finished. Just ran out of time today, so will push that tomorrow. I was thinking that needs to be a separate pull request as it is a (slightly) different issue?

**Comment by jamtur01** [comment]
> @jamtur01 Yes there is. I've almost got that finished. Just ran out of time today, so will push that tomorrow. I was thinking that needs to be a separate pull request as it is a (slightly) different issue?

Yeah - that sounds good. Was just checking on that.

File: src/sinks/datadog/logs.rs
Comment (fanatid) [inline src/sinks/datadog/logs.rs:222]: I think `match` will be nicer here

File: src/sinks/datadog/logs.rs
Comment (fanatid) [inline src/sinks/datadog/logs.rs:56]: Is not too much refs here? Should work:
```suggestion
        match &self.endpoint {
            Some(endpoint) => endpoint,
            None => "https://http-intake.logs.datadoghq.eu/v1/input",
        }
```

Or as alternative, can be changed to:
```rust
self.endpoint.as_ref().map(|s| s.as_str()).unwrap_or("https://http-intake.logs.datadoghq.eu/v1/input")
```

**Review by fanatid — APPROVED** [review]

**Comment by szibis** [comment]
Is that traffic is gzipped as Datadog suggest ?? https://docs.datadoghq.com/api/v1/logs/ 

> We encourage you to send your logs compressed. Add the Content-Encoding: gzip header to the request when sending compressed logs.

Also, can we add some enhancement as Datadog also have in Datadog-Agent - gzip compression level ?? https://docs.datadoghq.com/agent/logs/log_transport/?tab=https#log-compression

>   use_compression: true
  compression_level: 6

**Comment by FungusHumungus** [comment]
@szibis Good call. No it doesn't.

I will see what I can do.

**Comment by FungusHumungus** [comment]
Have just pushed a commit that adds compression to the sink.

I have put the compression directly into the sink rather than as a layer in the `BatchHttpSink`. It is the `Datadog Sink` that is responsible for creating the request and then passing it in to the `BatchHttpSink`. Theoretically, `BatchHttpSink` could then take this request apart and compress the body, but this didn't feel like the right solution. I was also unable to find anywhere closer to the Hyper services that had access to the Body stream, but it's possible there is something there I have missed. 

It was also not possible to put the compression in the Buffer. The Http Sink is able to compress through the Buffer because the buffer it uses already takes the raw bytes, so it can compress as it goes along. With the JsonArrayBuffer, we don't get the actual bytes until the end when the whole object is serialized. One downside to this is that the buffer is only aware of the bytes before compression to see how full it is. Obviously compression reduces this significantly. Is this going to be a problem?

I have tested this against Datadog, so it does work. I don't think what I have here is the best possible solution in terms of the architecture, as it is not reusable among all the Http sinks. But I'm not 100% sure of a better way at the minute. Would be interested in feedback.

I'll update the docs once we are confident of a way forward.

**Comment by binarylogic** [comment]
@bruceg should be able to chime in on that.

**Comment by bruceg** [comment]
As you note, any generic solution would have to happen between serializing the buffer into bytes (which is a no-op for the plain `Buffer` type) and turning that data into a request for transmission. I haven't reviewed this PR yet (will do that right away) but it sounds like the right approach for now. We'll consider a more general approach later.

I don't see another good way of doing it other than as you say — storing the records raw in the buffer and compressing when producing the transmission record. We could add the raw bytes to a compression buffer to estimate the compressed size, but that buffer would have to be thrown away at the end of the process, as it would be lacking the enclosing JSON bits to make it a proper record, so that would be doubling the CPU cost of compression.

File: src/sinks/datadog/logs.rs
Comment (bruceg) [inline src/sinks/datadog/logs.rs:46]: Note that the TLS settings are *not* obsolete here, as they can be used to control options other than turning on TLS mode — for example, certificates and keys.

File: src/sinks/datadog/logs.rs
Comment (bruceg) [inline src/sinks/datadog/logs.rs:81]: This appears to be the same in both branches. I suggest just unwrapping before the match.

File: src/sinks/datadog/logs.rs
Comment (bruceg) [inline src/sinks/datadog/logs.rs:89]: Would it make sense to do the indirection within `BatchedHttpSink` instead of outside? ie:
```
let sink = BatchedHttpSink::new(
    Box::new(service) as Box<dyn HttpSink>,
    Box::new(JsonArrayBuffer::new(batch_settings.size)) as Box<dyn Batch>,
```

File: src/sinks/datadog/logs.rs
Comment (bruceg) [inline src/sinks/datadog/logs.rs:184]: Since this takes `DatadogLogsConfig` as the first parameter, what do you think of moving this into a method of that struct?

File: src/sinks/datadog/logs.rs
Comment (bruceg) [inline src/sinks/datadog/logs.rs:230]: Perhaps a match?
```rust
    match status {
        StatusCode::OK => Ok(()),
etc
```

File: src/sinks/datadog/logs.rs
Comment (bruceg) [inline src/sinks/datadog/logs.rs:243]:
```suggestion
        Err(json
            .as_object()
            .and_then(|o| o.get("error"))
            .and_then(|s| s.as_str())
            .unwrap_or("Token is not valid, 401 returned.")
            .to_string()
            .into()
        )
```
Might have to flip the order of the last two lines if there's a lifetime conflict on the `&str`

File: src/sinks/datadog/logs.rs
Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs:81]: They look the same, but they infer to different types for each branch. For one `batchsettings` is `Batch<VecBuffer<_>>` and for the other branch it is  `Batch<JsonArrayBuffer>`.  It doesn't work before the match as it doesn't know which object to create (believe me, I tried!).

What I can do is move that out into a separate generic function so at least we don't have duplicate code. It's not ideal, if you can think of a better way, let me know!

File: src/sinks/datadog/logs.rs
Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs:184]: Seems logical.

File: src/sinks/datadog/logs.rs
Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs:89]: The issue I run into here is that `BatchedHttpSink` wants an `HttpSink` and not a `Box<HttpSink>`. Dereffing the Box means I have a `&dyn HttpSink`, which is also not what I want. I imagine if I did get around that it wouldn't be `Sized`, that could also cause further issues.

Unless you have any other ideas, it's possible I'm missing something?

I see ultimately `BatchedHttpSink` stores an `Arc<HttpSink>`, with a bit of rework we could perhaps add a new `new` method that allowed an `Arc` to be passed in instead. But I'm not convinced it would be worth it just for this one case.

Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs:89]: Monomorphisation to the rescue.. I can, however, extract most of the duplicated functionality out into a separate function that each branch can then call.

Comment (bruceg) [reply src/sinks/datadog/logs.rs]: No, no other ideas, was just curious about where best to do the boxing.

**Review by bruceg — CHANGES_REQUESTED** [review]
Code looks better, just a couple of questions from me. There are a couple of pieces missing from the docs still (TLS and compression settings), which will end up conflicting with the work @binarylogic is doing in #4303.

File: src/sinks/datadog/logs.rs
Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs:130]: Yes, I can see that being confusing. I just chose to default to the same level that was being run by the Http sink should an invalid level be entered.

Do you think it might be better to error here instead as if you specify a compression level of 10, you have probably made a mistake?

File: .meta/sinks/datadog_logs.toml.erb
Comment (FungusHumungus) [reply .meta/sinks/datadog_logs.toml.erb:54]: I think it should be pretty quick to sort out these doc changes, so it's probably worth spending the 10 minutes to do them rather than risk this pull sitting around for longer than necessary..

**Comment by binarylogic** [comment]
Yep, as @bruceg mentioned, I'd like to merge https://github.com/timberio/vector/pull/4350 first and then update that file accordingly.

File: src/sinks/datadog/logs.rs
Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs]: I have gone with just changing this to best. If we do want to validate the parameter, this will need to be done elsewhere anyway, so won't affect this code.

Comment (bruceg) [reply src/sinks/datadog/logs.rs]: Specifying an invalid compression level is probably worth a warning at least, and yes, validation would best be done elsewhere.

Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs]: Ok, I have moved the Compression Level into a newtype. I think it encapsulates the property better, although there is a fair amount of extra boiler plate. 

If you think it is a bit overkill, let me know and I can just put a quick check at the start of the `build` method  instead?

Comment (bruceg) [reply src/sinks/datadog/logs.rs]: It's a little extra boilerplate, but I don't think it's overkill. This will also be useful with support for compression levels in the generic buffer code, so it may end up used elsewhere as well.

File: src/sinks/datadog/logs.rs
Comment (fanatid) [inline src/sinks/datadog/logs.rs:79]: Can we remove `compression_level` field? Compression level should be added to Compression struct in https://github.com/timberio/vector/pull/3032.

File: src/sinks/datadog/logs.rs
Comment (fanatid) [inline src/sinks/datadog/logs.rs:267]:
```suggestion
        let body: Vec<u8> = events.into_iter().flat_map(Bytes::into_iter).collect();
```

File: .meta/sinks/datadog_logs.toml.erb
Comment (fanatid) [inline .meta/sinks/datadog_logs.toml.erb:1]: This file not required anymore.

File: src/sinks/datadog/logs.rs
Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs:79]: Ah nice. I hadn't seen that pull. That's a much better way to do it. How soon is that likely to be merged in?

Comment (fanatid) [reply src/sinks/datadog/logs.rs:79]: I hope that we will merge it on this week :slightly_smiling_face:

Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs:79]: Ok great. I'll hold off on this change until that lands.

File: docs/reference/components/sinks/datadog_logs.cue
Comment (binarylogic) [inline docs/reference/components/sinks/datadog_logs.cue:73]: Curious why you chose 6, is that the library default?

File: src/sinks/datadog/logs.rs
Comment (bruceg) [reply src/sinks/datadog/logs.rs:79]: #3032 was merged, so this can proceed.

**Review by bruceg — COMMENTED** [review]
Looks good except for the `Compression` level changes.

File: src/sinks/datadog/logs.rs
Comment (FungusHumungus) [reply src/sinks/datadog/logs.rs]: All done.

**Comment by binarylogic** [comment]
Nice work!

**Comment by binarylogic** [comment]
@FungusHumungus feel free to self-merge going forward once you get one or two approvals depending on the complexity of the change.

**Pull request merged**
